crates/bevy_asset/src/http_source.rs

```diff
@@ -0,0 +1,254 @@
+use crate::io::{AssetReader, AssetReaderError, Reader};
+use crate::io::{AssetSource, PathStream};
+use crate::AssetApp;
+use alloc::boxed::Box;
+use bevy_app::App;
+use bevy_tasks::ConditionalSendFuture;
+use std::path::{Path, PathBuf};
+
+/// Adds the `http` and `https` asset sources to the app.
+/// Any asset path that begins with `http` or `https` will be loaded from the web
+/// via `fetch`(wasm) or `ureq`(native).
+pub fn http_source_plugin(app: &mut App) {
+    app.register_asset_source(
+        "http",
+        AssetSource::build().with_reader(|| Box::new(HttpSourceAssetReader::Http)),
+    );
+    app.register_asset_source(
+        "https",
+        AssetSource::build().with_reader(|| Box::new(HttpSourceAssetReader::Https)),
+    );
+}
+
+/// Asset reader that treats paths as urls to load assets from.
+pub enum HttpSourceAssetReader {
+    /// Unencrypted connections.
+    Http,
+    /// Use TLS for setting up connections.
+    Https,
+}
+
+impl HttpSourceAssetReader {
+    fn make_uri(&self, path: &Path) -> PathBuf {
+        PathBuf::from(match self {
+            Self::Http => "http://",
+            Self::Https => "https://",
+        })
+        .join(path)
+    }
+
+    /// See [`crate::io::get_meta_path`]
+    fn make_meta_uri(&self, path: &Path) -> Option<PathBuf> {
+        let mut uri = self.make_uri(path);
+        let mut extension = path.extension()?.to_os_string();
```

Comment:
How and why can this fail?

If you have a file called `http://example.com/1234567890`, surely `http://example.com/1234567890.meta` is a reasonable path to look metadata in?